packages/fxa-auth-server/scripts/cleanup-partial-firestore-customers.ts

```diff
@@ -0,0 +1,231 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+/* eslint @typescript-eslint/no-unused-vars: 0 */
+import program from 'commander';
+
+import { setupProcessingTaskObjects } from '../lib/payments/processing-tasks-setup';
+import { parseDryRun } from './lib/args';
+import {
+  FieldPath,
+  Firestore,
+  QueryDocumentSnapshot,
+} from '@google-cloud/firestore';
+import {
+  AccountDeleteReasonsMap,
+  AppConfig,
+  AuthFirestore,
+  AuthLogger,
+} from '../lib/types';
+import Container from 'typedi';
+import { AccountDeleteManager } from '../lib/account-delete';
+import { ConfigType } from '../config';
+import oauthDb from '../lib/oauth/db';
+import { StatsD } from 'hot-shots';
+import pushboxApi from '../lib/pushbox';
+import { setupAccountDatabase } from '@fxa/shared/db/mysql/account';
+import { AccountManager } from '@fxa/shared/account/account';
+import { uuidTransformer } from 'packages/fxa-shared/db/transformers';
+import PQueue from 'p-queue';
+
+const pckg = require('../package.json');
+
+class CleanupFirestoreHelper {
+  private firestore: Firestore;
+  private log: AuthLogger;
+  private config: ConfigType;
+  private accountDeleteManager: AccountDeleteManager;
+  private accountManager: AccountManager;
+  private cloudTaskQueue: PQueue;
+
+  constructor(
+    private batchSize: number,
+    private dryRun: boolean,
+    rateLimit: number
+  ) {
+    this.firestore = Container.get<Firestore>(AuthFirestore);
+    this.log = Container.get(AuthLogger);
+    this.config = Container.get(AppConfig);
+    this.accountDeleteManager = Container.get(AccountDeleteManager);
+    this.accountManager = Container.get(AccountManager);
+    this.cloudTaskQueue = new PQueue({ concurrency: rateLimit });
+  }
+
+  /**
+   * Fetches subscriptions from Firestore paginated by batchSize
+   * @param startAfter ID of the last element of the previous batch for pagination
+   * @returns A list of subscriptions from firestore
+   */
+  private async fetchFirestoreCustomersUid(
+    startAfter: QueryDocumentSnapshot | undefined
+  ) {
+    this.log.debug('CleanupFirestoreHelper.fetchFirestoreCustomersUid.start', {
+      startAfterDocId: startAfter?.id,
+    });
+
+    const collectionPrefix = `${this.config.authFirestore.prefix}stripe-`;
+    const customerCollection = `${collectionPrefix}customers`;
+    const customerSnap = startAfter
+      ? await this.firestore
+          .collectionGroup(customerCollection)
+          .orderBy(FieldPath.documentId())
+          .startAfter(startAfter)
+          .limit(this.batchSize)
+          .get()
+      : await this.firestore
+          .collectionGroup(customerCollection)
+          .orderBy(FieldPath.documentId())
+          .limit(this.batchSize)
+          .get();
+
+    // Customer document ids are the FxA Account UID
+    const uids = customerSnap.docs.map((doc) => doc.id);
+    const newStartAfter = customerSnap.docs.at(-1);
+
+    return { uids, startAfter: newStartAfter };
+  }
```

Comment:
Might make for a simpler API to write a little async generator for this so the caller doesn't need to deal with tracking and passing a startAfter. For example:
```typescript
async *fetchFirestoreCustomersUids() {
  this.log.debug('CleanupFirestoreHelper.fetchFirestoreCustomersUid.start', {
      startAfterDocId: startAfter?.id,
    });
  const customerCollection = `${this.config.authFirestore.prefix}stripe-customers`;
  let customerSnap = await this.firestore
        .collectionGroup(customerCollection)
        .orderBy(FieldPath.documentId())
        .limit(this.batchSize)
        .get();
  while (!customerSnap.empty) {
    yield customerSnap.docs.map((doc) => doc.id);
    customerSnap = await this.firestore
          .collectionGroup(customerCollection)
          .orderBy(FieldPath.documentId())
          .startAfter(customerSnap.docs.at(-1))
          .limit(this.batchSize)
          .get();
  }
}
```

Then it can be used with:
```typescript
for await (const uids of this.fetchFirestoreCustomersUids()) {
  ...do stuff with uids...
}
```

Reply:
Oh very nice. Yup this works and looks much better. Thanks for the snippet.